IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,660 @@
+
+    if (items.count > 1)
+    {
+        [self createError:@"The token cache store for this resource contains more than one user"
+                   domain:MSIDErrorDomain errorCode:(NSInteger)MSIDErrorCacheMultipleUsers error:error context:context];
+        return nil;
+    }
+
+    return items.firstObject;
+}
+
+// Read one or more accounts from the keychain that match the key (see accountItem:matchesKey).
+// If not found, return an empty list without setting an error.
+//
+// Errors:
+// * MSIDKeychainErrorDomain/OSStatus: Apple status codes from SecItemCopyMatching()
+//
+- (NSArray<MSIDAccountCacheItem *> *)accountsWithKey:(MSIDCacheKey *)key
+                                          serializer:(id<MSIDAccountItemSerializer>)serializer
+                                             context:(id<MSIDRequestContext>)context
+                                               error:(NSError **)error
+{
+    MSID_TRACE;
+    MSID_LOG_VERBOSE(context, @"Get keychain items, key info (account: %@ service: %@ generic: %@ type: %@, keychainGroup: %@)", _PII_NULLIFY(key.account), key.service, _PII_NULLIFY(key.generic), key.type, [self keychainGroupLoggingName]);
+    MSID_LOG_VERBOSE_PII(context, @"Get keychain items, key info (account: %@ service: %@ generic: %@ type: %@, keychainGroup: %@)", key.account, key.service, key.generic, key.type, self.keychainGroup);
+
+    NSMutableDictionary *query = [self accountQueryForKey:key];
+    // Per Apple's docs, kSecReturnData can't be combined with kSecMatchLimitAll:
+    // https://developer.apple.com/documentation/security/1398306-secitemcopymatching?language=objc
+    // For this reason, we retrieve references to the items, then (below) use a second SecItemCopyMatching()
+    // to retrieve the data for each, deserializing each into an account object.
+    query[(id)kSecMatchLimit] = (id)kSecMatchLimitAll;
+    query[(id)kSecReturnRef] = @YES;
+
+    CFTypeRef cfItems = nil;
+    MSID_LOG_INFO(context, @"Trying to find keychain items...");
+    OSStatus status = SecItemCopyMatching((CFDictionaryRef)query, &cfItems);
+    MSID_LOG_INFO(context, @"Keychain find status: %d", (int)status);
+
+    if (status == errSecItemNotFound)
+    {
+        return @[];
+    }
+    else if (status != errSecSuccess)
+    {
+        [self createError:@"Failed to read account from keychain"
+                   domain:MSIDKeychainErrorDomain errorCode:(NSInteger)status error:error context:context];
+        return nil;
+    }
+
+    NSArray *items = CFBridgingRelease(cfItems);
+
+    query = [self accountQueryForKey:key];
+    // Note: For efficiency, use kSecUseItemList to query the items returned above rather actually querying
+    // the keychain again. With this second query we can set a specific kSecMatchLimit which lets us get the data
+    // objects.
+    query[(id)kSecUseItemList] = items;
+    query[(id)kSecMatchLimit] = @(items.count + 1); // always set a limit > 1 so we consistently get an NSArray result
```

Comment:
Just curious why is this +1?

Reply:
I found that if I just set it to items.count and items only has one element, the SecItemCopyMatching() will return a single item instead of an NSArray. Although a simple solution, it looks a little odd, as you've noted. It would probably be better to explicitly check for this and wrap it in an NSArray when this happens before proceeding to iterate across the array.